pkg/cache/context.go

```diff
@@ -454,6 +454,41 @@ func (ctx *Context) IsPodFitNode(name, node string, allocate bool) error {
 	return fmt.Errorf("predicates were not running because pod or node was not found in cache")
 }
 
+func (ctx *Context) IsPodFitNodeViaPreemption(name, node string, allocations []string, startIndex int) (index int, ok bool) {
+	// assume minimal pods need killing if running in testing mode
+	if ctx.apiProvider.IsTestingMode() {
+		return startIndex, ok
+	}
+
+	ctx.lock.RLock()
+	defer ctx.lock.RUnlock()
+	if pod, ok := ctx.schedulerCache.GetPod(name); ok {
+		// if pod exists in cache, try to run predicates
+		if targetNode := ctx.schedulerCache.GetNode(node); targetNode != nil {
+			// need to lock cache here as predicates need a stable view into the cache
+			ctx.schedulerCache.LockForReads()
+			defer ctx.schedulerCache.UnlockForReads()
+
+			// look up each victim in the scheduler cache
+			victims := make([]*v1.Pod, 0)
+			for _, uid := range allocations {
+				if victim, ok := ctx.schedulerCache.GetPod(uid); ok {
```

Comment:
`GetPod()` also acquires a read lock, this call blocks here.

Reply:
No, it doesn't. It takes the cache lock (which is a different lock). This mirrors the existing code in `IsPodFitNode()` which has been working reliably forever.

Reply:
Thanks, looks like I missed that.